src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceVariableService.cs

```diff
@@ -182,29 +185,45 @@ private async Task CreateConstantFieldActionsAsync(State state, ArrayBuilder<Cod
                 // If something is a constant, and it doesn't access any other locals constants,
                 // then we prefer to offer to generate a constant field instead of a constant
                 // local.
-                var action1 = CreateAction(state, allOccurrences: false, isConstant: true, isLocal: false, isQueryLocal: false);
-                if (await CanGenerateIntoContainerAsync(state, action1, cancellationToken).ConfigureAwait(false))
+                if (CanGenerateIntoContainer(state, cancellationToken))
                 {
-                    actions.Add(action1);
+                    actions.Add(CreateAction(state, allOccurrences: false, isConstant: true, isLocal: false, isQueryLocal: false));
+                    actions.Add(CreateAction(state, allOccurrences: true, isConstant: true, isLocal: false, isQueryLocal: false));
                 }
+            }
+        }
+
+        protected int GetFieldInsertionIndex(
+            bool isConstant, TTypeDeclarationSyntax oldType, TTypeDeclarationSyntax newType, CancellationToken cancellationToken)
+        {
+            var preferredInsertionIndex = isConstant
+                ? DetermineConstantInsertPosition(oldType, newType)
+                : DetermineFieldInsertPosition(oldType, newType);
+
+            var legalInsertionIndices = this.GetInsertionIndices(oldType, cancellationToken);
+            if (legalInsertionIndices[preferredInsertionIndex])
+            {
+                return preferredInsertionIndex;
+            }
 
-                var action2 = CreateAction(state, allOccurrences: true, isConstant: true, isLocal: false, isQueryLocal: false);
-                if (await CanGenerateIntoContainerAsync(state, action2, cancellationToken).ConfigureAwait(false))
+            // location we wanted to insert into isn't legal (i.e. it's hidden).  Try to find a
+            // non-hidden location.
+            for (var i = 0; i < legalInsertionIndices.Count; i++)
+            {
+                if (legalInsertionIndices[i])
                 {
-                    actions.Add(action2);
+                    return i;
                 }
             }
+
+            // Couldn't find a viable non-hidden position.  Fall back to the computed position we
+            // wanted originally.
+            return preferredInsertionIndex;
```

Comment:
this is actually nicer behavior than before.  previously, we might actually insert into a hidden location (if choosing 'introduce field').  Now, we actually try to find a non-hidden location if it exists first.  And only if it doesn't do we insert into a hidden location.